### ISSUE TITLE
Recommended way to check for sparse data (DataFrame or Series)

### ISSUE THREAD

**Issue by jorisvandenbossche** [pandas-dev/pandas#26706]
Previously, you check for the SparseDataFrame class, but now we want people to store sparse data in normal DataFrames/Series. 

So for a library developer, what is the recommended way to check for sparse data?

For a Series (or SparseArray), the `pd.api.types.is_sparse` still works on a plain Series with sparse data:

```
In [42]: df = pd.DataFrame({'a': pd.SparseArray([1, np.nan, 1])}) 

In [43]: s = df['a']

In [44]: type(s)
Out[44]: pandas.core.series.Series

In [45]: pd.api.types.is_sparse(s)
Out[45]: True
```

but for a DataFrame that function does not work. So you could apply that method on each column/dtype, and check with `any` or `all` depending on your requirements:

```
In [56]: df = pd.DataFrame({'a': pd.SparseArray([1, np.nan, 1]), 'b': [1, 2, 3]})

In [57]: df.dtypes.apply(pd.api.types.is_sparse)
Out[57]: 
a     True
b    False
dtype: bool

In [58]: df.dtypes.apply(pd.api.types.is_sparse).any()
Out[58]: True

In [59]: df.dtypes.apply(pd.api.types.is_sparse).all()
Out[59]: False
```

So that actually works quite OK, now I am writing it. 
Do we want to make this even easier somehow? Or document this?

**Comment by TomAugspurger** [comment]
A helper pd.DataFrame.sparse.is_sparse -> Series[bool] seems reasonable.
The index would be the column names.

**Comment by jorisvandenbossche** [comment]
Ah, yes, that might be a good idea.

**Comment by jreback** [comment]
no i wouldn’t adds special method at all

rather we already have select_dtypes which should.(and might) accept ‘sparse’ or more specific as well

**Comment by TomAugspurger** [comment]
select_dtypes is a bit different from a boolean mask indicating whether a
column is sparse or not.

**Comment by marctorrellas** [comment]
I agree, having a `.is_sparse` in each pd.Series would be helpful. Sometimes you just want to know which ones are sparse, not to create a new df with them

**Comment by jorisvandenbossche** [comment]
PR always welcome!

**Comment by SdgJlbl** [comment]
I'll work on that

**Comment by jorisvandenbossche** [comment]
> [@jreback] we already have multiple ways of selectimg dtypes and i don’t see any justification

Again, this is not necessarily about selecting dtypes, it is about knowing if you have a sparse dataframe or not.

> even in the accessor this doesn’t make any sense (even less sense)

Can you to explain *why* you think this makes even less sense in the accessor? It's true that you can call this on non-sparse data, but it still is related to the sparse functionality.

**Comment by jreback** [comment]
instead of me explaining an objection how about u justify with an example that you cannot do now?

**Comment by jorisvandenbossche** [comment]
The `pd.api.types` functions are rather hidden and due to their nested access seemingly rather for library developers (+ needing to apply the function on the dtypes is also not something straightforward), while wanting to know that your dataframe is sparse, seems to be something rather typical to want to check.  
Therefore, I think that is method on the sparse accessor is a nice alternative to `df.dtypes.apply(pd.api.types.is_sparse)`.

**Comment by jorisvandenbossche** [comment]
@jreback do you have further objections?

**Comment by jorisvandenbossche** [comment]
@jreback ping. Otherwise I plan to go forward with the PR, multiple people here seem to find this a useful addition.

**Comment by jreback** [comment]
i will comment later and yes i still object

**Comment by jreback** [comment]
We already have the ``pd.api.types``,  ``.select_dtypes``, and ``.dtypes``. So you are proposing yet another way to do virtually the same thing.

You are also essentially proposing a massive api expansion of ``is_*`` on Series (and Index) for that matter (this is at least 10 properties).

I find this wildly confusing and am -1 on this addition.

If you thing this is the 'best' api, then please propose a unification & deprecation plan first.

**Comment by jorisvandenbossche** [comment]
> You are also essentially proposing a massive api expansion of is_* on Series (and Index) for that matter (this is at least 10 properties).

No, I never proposed that. I only proposed a single property on the `sparse` accessor for a DataFrame (so not even top-level on DataFrame), and nothing for Index / Series. 
Where did you read this massive API expansion? (trying to understand our misunderstanding)

**Comment by jorisvandenbossche** [comment]
@jreback Can you respond to my last answer? 

So to be clear: this is about a single property (or method) in the `sparse` accessor (which is what the PR is doing), nothing more than that is proposed.

**Comment by jreback** [comment]
> @jreback Can you respond to my last answer?
> 
> So to be clear: this is about a single property (or method) in the `sparse` accessor (which is what the PR is doing), nothing more than that is proposed.

I get what you are proposing, but nothing has changed my mind. This is a scope creep that won't end well. I don't see what is wrong with an apply here. This seems like api that is wanting for a usecase.

**Comment by jorisvandenbossche** [comment]
> This seems like api that is wanting for a usecase.

Use case: knowing if you ave only sparse columns (some functionality will only work in such case). Before, you could do this with SparseDataFrame, but now in a normal DataFrame you cannot make any guarantees about having only sparse columns.

**Comment by jorisvandenbossche** [comment]
Example application: scikit-learn wants to know if a passed DataFrame is sparse or not (https://github.com/scikit-learn/scikit-learn/issues/15976)

**Comment by TomAugspurger** [comment]
@jorisvandenbossche moving to contributions welcome, unless you think we'll get to it for 1.0 (can be in the RC period).

**Comment by SdgJlbl** [comment]
At the time I had started a PR on this issue (#28314), should it be revived?

**Comment by jorisvandenbossche** [comment]
@TomAugspurger it's an easy thing to add, we just need to decide if we want to add it. I think there are good use cases, and as far as I know, there are several people +1 on adding it, only @jreback is -1

**Comment by TomAugspurger** [comment]
Gotcha. IMO

* I'm +1 on adding this
* It's not a blocker for 1.0
* It can be done during the RC

**Comment by jreback** [comment]
-1 this uneeded api bloat 

this smells of a SparseDataftame again who is a terrible idea

**Comment by jreback** [comment]
i would not object to a top level function say 

api.types.is_all_sparse or something

**Comment by jorisvandenbossche** [comment]
How does this "smell of a SparseDataFrame" ? 
Please, if you are so categorically -1, try to explain better why.

Whether you like it or not, but we have a sparse dtype, we have a `.sparse` accessor, and we have some sparse related methods (eg to convert to scipy sparse matrix). So having a helper method to be able to check if you have sparse data and thus can call those sparse methods, seems rather logical to me. Scikit-learn is now doing something like `hasattr(df, 'sparse')`, which would be avoided by this.

**Comment by jreback** [comment]
> How does this "smell of a SparseDataFrame" ?
> Please, if you are so categorically -1, try to explain better why.
> 
> Whether you like it or not, but we have a sparse dtype, we have a `.sparse` accessor, and we have some sparse related methods (eg to convert to scipy sparse matrix). So having a helper method to be able to check if you have sparse data and thus can call those sparse methods, seems rather logical to me. Scikit-learn is now doing something like `hasattr(df, 'sparse')`, which would be avoided by this.

You need to provide a very good justification for this. I don't think your reasons are good enough. We *already* have good ways of testing for sparseness, we DON'T need another. I am categorically -1. I have already explained why multiple times and offered how you can do it.

**Comment by marctorrellas** [comment]
just to reiterate, I'm +1. I don't think it's anything redundant. Obviously you can do this with an apply().all(), but there are other things in pandas you can do in some other way e.g. values vs to_numpy(). The addition of this, IMHO and as explained above, makes this task more intuitive to guess and readable to review external code

**Comment by TomAugspurger** [comment]
Doesn't seem to be happening for 1.0. Pushing.

**Comment by jorisvandenbossche** [comment]
Ran into this again (https://github.com/scikit-learn/scikit-learn/pull/16728), scikit-learn starts to use `hasattr(df, 'sparse')` as the way to check that all columns are sparse. IMO that is really not a good and robust way. Having a `df.sparse.is_sparse.any()/all()` would make this clearer and more robust.

**Comment by TomAugspurger** [comment]
Yeah, as you commented there that's definitely not robust. This is a decent argument for having a `.sparse.is_sparse`, as an indication that `.sparse` is a DataFrame accessor that may be present on mixed / all-dense DataFrames in the future.

**Comment by marctorrellas** [comment]
Excuse my ignorance, but why do we need `df.sparse.is_sparse.all()` when we could have `df.is_sparse.all()`? Don't we have `df.isnull().all()` ? I understand the latter works field by field, whereas `is_sparse` would work column by column, but still, I don't find a reason to write the word `sparse` twice

**Comment by jorisvandenbossche** [comment]
There is no "need" for that from a technical point of view. But rather, it is to avoid adding a lot of methods/attributes directly on the DataFrame object, so we have accessors that group some data-type specific functionality (like `str`, `dt`, `cat` on Series). 
For that reason, I proposed to have it under the `df.sparse` accessor (since it already exists), but it *could* also be directly on the DataFrame.

**Comment by avinashpancham** [comment]
take

**Comment by avinashpancham** [comment]
Not sure whether we wanted to implement this or not, but I gave it a try. Feel free to close the PR if we don't want to continue with this.

**Comment by jbrockmendel** [comment]
Is `df.dtypes.map(lambda x: isinstance(x, pd.SparseDtype))` too inconvenient?

**Comment by jbrockmendel** [comment]
@rhshadrach agree with me that `df.dtypes.map(lambda x: isinstance(x, pd.SparseDtype))` gets the job done?

**Comment by rhshadrach** [comment]
That seems reasonable to me, but I'd like @jorisvandenbossche's thoughts here too before closing.

**Comment by jbrockmendel** [comment]
Closing as no action.
